source/aws_http_credentials_provider.c

```diff
@@ -0,0 +1,515 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/auth/credentials.h>
+#include <aws/auth/private/aws_http_credentials_provider.h>
+#include <aws/auth/private/credentials_utils.h>
+#include <aws/common/clock.h>
+#include <aws/http/request_response.h>
+#include <aws/http/status_code.h>
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+
+#define HTTP_RESPONSE_SIZE_INITIAL 2048
+#define HTTP_RESPONSE_SIZE_LIMIT 10000
+#define HTTP_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS 2
+#define HTTP_MAX_ATTEMPTS 3
+#define HTTP_RETRY_TIMEOUT_MS 100
+
+struct aws_credentials_provider_http_impl {
+    struct aws_http_connection_manager *connection_manager;
+    const struct aws_auth_http_system_vtable *function_table;
+    struct aws_string *endpoint;
+    struct aws_retry_strategy *retry_strategy;
+    struct aws_http_credentials_provider_request_vtable *request_vtable;
+};
+
+static void s_http_query_context_reset_request_specific_data(struct aws_http_query_context *http_query_context) {
+    struct aws_credentials_provider_http_impl *impl = http_query_context->provider->impl;
+    impl->request_vtable->destroy_request_data_fn(http_query_context->request_data);
+    if (http_query_context->request) {
+        aws_http_message_release(http_query_context->request);
+        http_query_context->request = NULL;
+    }
+    if (http_query_context->connection) {
+        int result = impl->function_table->aws_http_connection_manager_release_connection(
+            impl->connection_manager, http_query_context->connection);
+        (void)result;
+        AWS_ASSERT(result == AWS_OP_SUCCESS);
```

Comment:
asserts are debug only and of rather limited use. (not sure we even run debug ci in this repo). 
is this an error that can happen in theory and is it catastrophic? can we bubble it up?

Reply:
[not a great reason but this was how it was designed before](https://github.com/awslabs/aws-c-auth/blob/main/source/credentials_provider_sso.c#L71-L78). for the scope of this i dont want to change this behavior, but also would if we have a reason to change it right now.